body.go

```diff
@@ -0,0 +1,64 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+)
+
+type body struct {
```

Comment:
I'm not entirely sure about the visibility.  If I expose a NewBody method that returns body, and I want that returned body to be opaque, do I still need to export it?

I was also thinking that the body struct would be a purely internal function, with only the `Download()` and `Upload()` methods exposed.

Reply:
If calling code needs a handle on the body (even if it is opaque) it will need to be exported, otherwise it won't be able to pass it back to e.g. another function call that receives it, and it won't be possible to call other methods on it. If it is a struct, you can "hide" the individual members by not exporting them. Then anything in your package can use these struct members, but no code outside sees the members or can do anything with them.